test/topology_experimental_raft/test_topology_recovery_basic.py

```diff
@@ -0,0 +1,77 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+import asyncio
+import logging
+import pytest
+import time
+
+from test.pylib.manager_client import ManagerClient
+from test.pylib.util import wait_for_cql_and_get_hosts
+from test.topology.util import reconnect_driver, restart, enter_recovery_state, \
+        delete_raft_data_and_upgrade_state, log_run_time, wait_until_upgrade_finishes as wait_until_schema_upgrade_finishes, \
+        wait_until_topology_upgrade_finishes, delete_raft_topology_state, check_system_topology_and_cdc_generations_v3_consistency
+
+
+@pytest.mark.asyncio
```

Comment:
Please run the new tests 100 times in debug mode

Reply:
They are currently running here: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2059/ . Will check the results tomorrow.

Reply:
Sigh, there was one crash.

A test was setting a 3-node cluster and, when the third node was being added, the second one crashed:

```
INFO  2024-02-08 21:47:54,586 [shard 0:strm] cdc - Started using generation (2024/02/08 19:47:54, f363ae60-c6ba-11ee-92a9-279f76963638).
DEBUG 2024-02-08 21:47:54,605 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain is called
DEBUG 2024-02-08 21:47:54,606 [shard 0:strm] raft_topology - reload raft topology state
DEBUG 2024-02-08 21:47:54,624 [shard 0:strm] raft_topology - topology_state_load: current CDC generation ID: (2024/02/08 19:47:54, f363ae60-c6ba-11ee-92a9-279f76963638)
DEBUG 2024-02-08 21:47:54,625 [shard 0: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 6, current version 6
DEBUG 2024-02-08 21:47:54,625 [shard 0: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-02-08 21:47:54,626 [shard 1: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 6, current version 6
DEBUG 2024-02-08 21:47:54,626 [shard 1: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-02-08 21:47:54,658 [shard 0:strm] raft_topology - reload raft topology state
DEBUG 2024-02-08 21:47:54,682 [shard 0:strm] raft_topology - topology_state_load: current CDC generation ID: (2024/02/08 19:47:54, f363ae60-c6ba-11ee-92a9-279f76963638)
ERROR 2024-02-08 21:47:54,709 [shard 1:main] database - regular column id 0 >= 0 ...
```

I see there is an issue already which, superficially, looks similar: https://github.com/scylladb/scylladb/issues/16821 . There, the crash also happens when adding a node in raft topology mode, the error message is similar (most likely crashes in the same place) - however it happens during repair, and here... I don't know where it happened, there is a backtrace but it's not decoded.

Will investigate further tomorrow (for real this time).

Reply:
Yeah, I'm pretty sure that it's #16821.

- The test is known to be flaky
- The failure happens at the beginning of the test, when a cluster in raft mode is being set up. The PR slightly changes how new nodes join the cluster (they do raft discovery earlier and there is an additional RPC to ask the cluster about which mode to use), but judging by the logs it looks like it works correctly. The crash only happens during streaming during bootstrap, which should not be affected by this PR.